Update information for translating the interface with Crowdin (#16718)
nvaccess/nvda: projectDocs/translating/crowdin.md

# Translating using Crowdin

Crowdin is used to translate the main NVDA interface.
NVDA's Crowdin project: <https://crowdin.com/project/nvda>.

This document covers setting up a Crowdin account, connecting it with PoEdit, and translating the main interface using Crowdin and PoEdit.

## Setup

### Create Crowdin account

1. Create a [Crowdin account](https://accounts.crowdin.com/register?continue=%2Fproject%2Fnvda).
1. Message the [translators mailing list](https://groups.io/g/nvda-translations) or <info@nvaccess.org> to request being added as a translator.
Please include your Crowdin username and the languages you wish to translate.

#### Translation reviews

> Translated strings will need to be reviewed and approved by a proofreader before being included in NVDA.

Reviewer: If I am not mistaken, the job of the approver is not covered in this document. I think that you cannot synchronize poedit with Crowdin to approve strings; the only thing that you can do is to upload your .po file and check auto-approval when doing this operation. To be checked though.

But explaining how strings should be approved is required, especially for people who start translating a new language. Else, this document is useless for new translators.

Author: I've added more information here

Translators may have proofreader status granted on a case-by-case basis by messaging the above contact points.

### Setup PoEdit

It is recommended that you use the latest version of PoEdit and NVDA for translating.
Alternatively, you can use Crowdin's interface directly.

PoEdit supports connecting with Crowdin directly.
PoEdit's homepage is: <http://www.poedit.net/>

1. Download the latest Windows PoEdit version at <https://poedit.net/download>
1. Install it by following the on-screen instructions, the default options should be sufficient.
1. When launching PoEdit:
    1. Choose "Translate cloud project"
    1. Connect your Crowdin account
    1. Select NVDA and the language you wish to translate

> ## Translating using PoEdit

Reviewer: The following information is missing and should be added in this paragraph:
* some word have plural forms; describe how to manage it, especially with poedit in which object navigation is required to reach the plural form
* some string have context information (when `pgettext` is used in the code)

Author: I've added new sections here, let me know what you think.

If you are unsure of meaning of the original interface message, consult automatic comments (also called translator comments).
Some comments provide an example output message to help you understand what NVDA will say when speaking or brailling such messages.

## Translating the interface

Open "nvda.po" for the language you want to translate in PoEdit.
Alternatively, you can use Crowdin's interface directly.

### Messages with formatting strings

You will come across several messages that have additional characters or punctuation as part of the message, this section will explain how they should be treated.

- `%d`, digits
- `%s`, string
- `%.2f`, a number to 2 digits after the comma, for example 5.25
- `{keyword}`, the text around the keyword should be translated, but the word should be left as is.

#### Examples

- `%d percent`: this means that `%d` will be replaced by a number when the program is running, and you only need to translate the word `percent`.
- `subject: %s`: here `%s` means that another string will be substituted.

> - `on {backgroundColor}`: In this case, the word on should be translated, and the rest should be left alone.

Reviewer: With the real string in `nvda.po`, matching the example:
```suggestion
- `{color} on {backgroundColor}`: In this case, the word on should be translated, and the rest should be left alone.
```

or even with the word "on" formatted:
```suggestion
- `{color} on {backgroundColor}`: In this case, the word `on` should be translated, and the rest should be left alone.
```

Author: ```suggestion
- `{color} on {backgroundColor}`: In this case, the word `on` should be translated.
The rest should be left alone or rearranged in order to suit the language.
```

Alternatively, you can use the [Crowdin web interface directly(https://support.crowdin.com/online-editor/).
As PoEdit only supports viewing approved strings, large translators team need to co-ordinate submitting unapproved strings to prevent conflicts.
Using Crowdin's interface avoids this problem.

PoEdit supports connecting with Crowdin directly.
PoEdit's homepage is: <http://www.poedit.net/>

1. Download the latest Windows PoEdit version at <https://poedit.net/download>
1. Install it by following the on-screen instructions, the default options should be sufficient.
1. When launching PoEdit:
    1. Choose "Translate cloud project"
    1. Connect your Crowdin account
    1. Select NVDA and the language you wish to translate

### Translation reviews

Translated strings will need to be reviewed and approved by a proofreader before being included in NVDA.
A proofreader is required for each language.
Proofreader status is granted on a case-by-case basis by messaging the [translators mailing list](https://groups.io/g/nvda-translations) or <info@nvaccess.org>

Proofreaders approve strings using the [Crowdin web interface](https://support.crowdin.com/online-editor/).

> PoEdit only supports viewing approved strings.

Reviewer: To be checked but, it seems to me, more precisely, that when you use synchronization of poedit with Crowdin, you get the approved strings from Crowdin, but of course, your personal changes (suggestions) also appear in poedit.

Author: ```suggestion
PoEdit does not support viewing unapproved strings from other translators.
```

> When submitting to Crowdin from PoEdit, proofreaders are able to auto-approve all submitted strings.

Reviewer: More specifically, you need to manually upload the file on Crowdin's web interface and check an auto-approve checkbox when doing this. But if I am a proof reader and unique translator of a language, I cannot just translate in poedit and hope that synchronization will allow to submit my work with auto-approval unfortunately.
Or is there a better way to do things?

Author: ```suggestion
When manually uploading to Crowdin from PoEdit, proofreaders are able to auto-approve all submitted strings.
```

A fuzzy string is a message which has been automatically translated, thus it may be wrong.
PoEdit will collect the new and fuzzy messages and presents them at the top of the already translated messages.

To insert or correct the translation for a string, first select it with the arrows, then tab to the blank edit field and type its translation.

NVDA will beep if you are on an untranslated or fuzzy message.
If you are using a braille display you'll see a star sign in-front of the messages you have to translate.

You may want to spell the original string to be aware of any punctuation mark, capital letters, etc.
PoEdit has a keystroke you may press while on an original string, `alt+c`, that copies the original string to the edit field when pressed.
You may then replace it with your translation as normal.

Press `control+s` at any moment to save your work.
Each time you press this key, PoEdit saves the po file, and if you check compile mo file checkbox in preferences, the .mo file will be re-generated.

NVDA provides additional shortcuts for PoEdit which are described in [the User Guide](https://www.nvaccess.org/files/nvda/documentation/userGuide.html#Poedit).

> If you are unsure of meaning of the original interface message, consult automatic comments (also called translator comments).

Reviewer: ```suggestion
If you are unsure of meaning of the original interface message, consult automatic comments (also called translator comments) pressing `control+shift+a`.
```

  - Few: `تتضمَّن %s عناصر`
  - Many: `تتضمَّن %s من العناصر`
  - Other: `تتضمَّن %s من العناصر`

In the [Crowdin web interface](https://support.crowdin.com/online-editor/), these can be set for each language using the "Form" section which replaces the standard translation edit box.

In PoEdit, the standard translation edit box has tabs for each plural form.
[Object navigation](https://www.nvaccess.org/files/nvda/documentation/userGuide.html#ObjectNavigation) is required to move focus to each tab button for the plural form.
You can do this by:

- By moving to the previous focus object from the edit box, you can cycle through each plural form tab button by continuing to more backward.
  - Desktop: `NVDA+numpad4`
  - Laptop: `NVDA+shift+leftArrow`
- Activate the current object once the desired tab button is reached.
  - Desktop: `NVDA+numpadEnter`
  - Laptop: `NVDA+enter`

If the number of plural forms for your language is incorrect please message the [translators mailing list](https://groups.io/g/nvda-translations) or <info@nvaccess.org>.

### String groupings

Translators may wish to provide two different translations for the same English string depending on the context.

> Translation strings can be grouped by a tag to differentiate contexts.

Reviewer: Wouldn't "with" be better here?

```suggestion
Translation strings can be grouped with a tag to differentiate contexts.
```

Author: I think they are equivalent, no?

Author: @coderabbitai which grammar is more correct here? "grouped by" or "grouped with"?

Author: interesting, learn some new thing every day huh

Reviewer: The Code Rabbit AI wrote:
> @seanbudd In the context provided, "grouped by a tag" is more grammatically appropriate. This phrasing clearly indicates that the tag is the criterion used for grouping, which aligns well with the intention of differentiating translation strings based on specific contexts. "Grouped with a tag" could imply that the tag is merely accompanying the strings rather than serving as the basis for grouping.

@seanbudd, the rabbit is correct in explaining both sense but cannot guess what you wanted to write. Maybe it was neither clear to me what you wanted to write.

But "grouped with tag" is more representative of my personal experience. That is, the tag appears along with the string to translate to provide more context (as the translators comment also does).

I am not sure at all that all the strings with context "addonStore" are following each other in the list of strings to translate, making a group of strings related one with another (i.e. "grouped by tage"): at least, it is not the case when untranslated strings are grouped together, either in poedit or on Crowdin's web interface.

It's not a matter of grammar correctness but of meaning.

Author: I think you are interpreting grouping as being in the same physical group in all list views, as opposed to contextual grouping. These tags are logically grouped based on context, for example, you can search by tag in PoEdit and Crowdin.

Reviewer: Yes, I was rather interpreting "grouping" as physical grouping rather than contextual grouping.

Anyway, my personal impression is that there is no specific grouping, i.e. "addonStore" strings are not gathered logically. I feel more that the context is an additional information / restriction indicating in which case this translation applies. To summarize, I would not use  "grouping" at all, but rather speak of tags and context only.

Though, it's really a matter of personal and subjective feeling. Feel free to ignore my comments on this topic, the more if no one expresses the same feeling!

> This allows differentiating the string "Status" being used in the Add-on Store context, as opposed to a different context.

Author: ```suggestion
```

Translator comments provide additional context information.

> Example: `addonStore` for `Status`.

Reviewer: Let's use a more illustrative example:

```suggestion
The string `none` is defined three times:
* without any context to report a lack of background pattern in Microsoft Excel
* with the context tag `symbolLevel`
* with the context tag `espeakVarient`
```

# Translating using Crowdin

Crowdin is used to translate the main NVDA interface.
NVDA's Crowdin project: <https://crowdin.com/project/nvda>.

This document covers setting up a Crowdin account, connecting it with PoEdit, and translating the main interface using Crowdin and PoEdit.

## Setup

### Create Crowdin account

1. Create a [Crowdin account](https://accounts.crowdin.com/register?continue=%2Fproject%2Fnvda).
1. Message the [translators mailing list](https://groups.io/g/nvda-translations) or <info@nvaccess.org> to request being added as a translator.
Please include your Crowdin username and the languages you wish to translate.

### Setup PoEdit

It is recommended that you use the latest version of PoEdit and NVDA for translating.

> Alternatively, you can use the [Crowdin web interface directly(https://support.crowdin.com/online-editor/).

Reviewer: Fix link:
```suggestion
Alternatively, you can use the [Crowdin web interface](https://support.crowdin.com/online-editor/) directly.
```
